File: Watch Match WatchKit Extension/InterfaceController.m
//
//  InterfaceController.m
//  Watch Match WatchKit Extension
//
//  Created by Watch Set Team on 11/23/14.
//  Copyright (c) 2014 Watch Set Team. All rights reserved.
//

#import "InterfaceController.h"

#import "GameLogic.h"
#import "GameTurn.h"
#import "PlayerIndex.h"

static const NSTimeInterval kTurnSeconds = 3.5;

@interface InterfaceController()
@property (weak, nonatomic) IBOutlet WKInterfaceLabel *gameStartsAtLabel;
@property (weak, nonatomic) IBOutlet WKInterfaceLabel *startTimeLabel;
@property (weak, nonatomic) IBOutlet WKInterfaceDate *startTimeDate;
@property (weak, nonatomic) IBOutlet WKInterfaceTimer *startTimeTimer;
@property (weak, nonatomic) IBOutlet WKInterfaceButton *button;
@property (weak, nonatomic) IBOutlet WKInterfaceImage *image;

@property (nonatomic) NSTimer *timer;

@property (nonatomic) GameLogic *gameLogic;

@property (nonatomic) NSUInteger turnIndex;

@property (nonatomic) BOOL isPlaying;

@property (nonatomic) NSMutableArray *tapTimes;

@property (nonatomic) NSMutableArray *didTapForTurns;

@property (nonatomic) NSUInteger hits;
@property (nonatomic) NSUInteger misses;

@end


@implementation InterfaceController

- (instancetype)initWithContext:(id)context {
    self = [super initWithContext:context];

    return self;
}

- (void)willActivate {
    self.gameLogic = [[GameLogic alloc] init];
    
    self.startTimeLabel.text =
    [NSDateFormatter localizedStringFromDate:self.gameLogic.startTime dateStyle:NSDateFormatterNoStyle timeStyle:NSDateFormatterShortStyle];
    
    self.timer = [[NSTimer alloc] initWithFireDate:self.gameLogic.startTime interval:0 target:self selector:@selector(startGameTimerDidFire:) userInfo:nil repeats:NO];
    
    self.title = @"Hits:";
    
    self.tapTimes = [[NSMutableArray alloc] init];
    
    self.didTapForTurns = [[NSMutableArray alloc] init];
    
    for (NSUInteger i = 0; i < self.gameLogic.turns.count; i++) {
        [self.didTapForTurns addObject:@(0)];
    }
    
    [[NSRunLoop mainRunLoop] addTimer:self.timer forMode:NSDefaultRunLoopMode];
}

- (IBAction)tapButton:(id)sender {
    if (self.isPlaying) {
        GameTurn *turn = self.gameLogic.turns[self.turnIndex];
        if (turn.isMatch) {
            if ([self.didTapForTurns[self.turnIndex] boolValue] == NO) {
                [self.tapTimes addObject:[NSDate date]];
                
                self.didTapForTurns[self.turnIndex] = @(YES);
                self.hits++;
                
                [self setTitle:[NSString stringWithFormat:@"Hits: %lu", (unsigned long)self.hits]];
            }
        } else {
            [self.tapTimes addObject:[NSDate date]];
        }
    }
}

- (void)showGame {
    self.gameStartsAtLabel.hidden = YES;
    self.startTimeLabel.hidden = YES;
    self.startTimeDate.hidden = YES;
    self.startTimeTimer.hidden = YES;
    self.button.color = [UIColor greenColor];
    self.isPlaying = YES;
}

- (void)startGameTimerDidFire:(NSTimer *)timer {
    [self showGame];
    
    [self playTurn];
    
    self.timer = [NSTimer scheduledTimerWithTimeInterval:kTurnSeconds target:self selector:@selector(playGameTimerDidFire:) userInfo:nil repeats:YES];
}

- (void)playGameTimerDidFire:(NSTimer *)timer {
    self.turnIndex++;
    
    if (self.turnIndex == self.gameLogic.turns.count) {
        [timer invalidate];
        self.timer = nil;
        [self endGame];
    } else {
        [self playTurn];
    }
}

- (void)playTurn {
    GameTurn *turn = self.gameLogic.turns[self.turnIndex];
    
    const NSUInteger shape = [turn shapeForPlayer:kPlayerIndex];
    const NSUInteger pattern = [turn patternForPlayer:kPlayerIndex];
    
    NSString *imageFileName = [NSString stringWithFormat:@"shape%lupattern%lu", (unsigned long)shape, (unsigned long)pattern];
    
    [self.image setImageNamed:imageFileName];
    [self.button setBackgroundImageNamed:imageFileName];
}

- (void)endGame {
    NSURL *libraryURL = [[NSFileManager defaultManager] URLForDirectory:NSApplicationSupportDirectory inDomain:NSUserDomainMask appropriateForURL:nil create:YES error:nil];
    NSURL *fileURL = [libraryURL URLByAppendingPathComponent:@"Taps.plist"];
    
    
    NSError *error;
    
    NSData *data = [NSPropertyListSerialization dataWithPropertyList:self.tapTimes format:NSPropertyListXMLFormat_v1_0 options:0 error:&error];
    
    BOOL result = [data writeToURL:fileURL options:0 error:&error];
    if (!result) {
        NSLog(@"%@", error);
    }
}

@end

File: Watch Match WatchKit Extension/InterfaceController.swift
//
//  InterfaceController.swift
//  Watch Match WatchKit Extension
//
//  Created by Andrew Pontious on 11/26/14.
//  Copyright (c) 2014 TeamHand. All rights reserved.
//

import WatchKit
import Foundation

import WatchFramework

class InterfaceController: WKInterfaceController {
    
    // Currently hard-coded, must be edited for every player in demo. Eventually: determined by initiating GameKit player.
    let kPlayerIndex = 0;

    let kTurnSeconds: NSTimeInterval = 3.5
    
    @IBOutlet weak var gameStartsAtLabel: WKInterfaceLabel!
    @IBOutlet weak var startTimeLabel: WKInterfaceLabel!
    @IBOutlet weak var button: WKInterfaceButton!
    @IBOutlet weak var gameOverLabel: WKInterfaceLabel!
    
    var timer: NSTimer?
    
    var gameLogic: GameLogic
    
    var turnIndex: Int
    
    var isPlaying: Bool
    
    var tapTimes: [NSDate]
    
    var didTapForTurns: [Bool]
    
    var hits: Int
    var misses: Int

    override init(context: AnyObject?) {
        self.gameLogic = GameLogic()
        self.turnIndex = 0
        self.isPlaying = false
        self.tapTimes = [NSDate]()
        self.didTapForTurns = [Bool]()
        self.hits = 0
        self.misses = 0
        
        super.init(context: context)
    }

    override func willActivate() {
        super.willActivate()
        
        self.startTimeLabel.setText(NSDateFormatter.localizedStringFromDate(self.gameLogic.startTime, dateStyle:NSDateFormatterStyle.NoStyle, timeStyle:NSDateFormatterStyle.ShortStyle));

        self.timer = NSTimer(fireDate:self.gameLogic.startTime,
                             interval:0,
                             target:self,
                             selector:Selector("startGameTimerDidFire:"),
                             userInfo:nil,
                             repeats:false)
        
        self.setTitle("Hits:")
        
        for i in 0..<self.gameLogic.turns.count {
            self.didTapForTurns.append(false)
        }
        
        NSRunLoop.mainRunLoop().addTimer(self.timer!, forMode:NSDefaultRunLoopMode)
    }

    @IBAction func tapButton() {
        if (self.isPlaying) {
            let turn = self.gameLogic.turns[self.turnIndex]
            if (turn.isMatch) {
                if (self.didTapForTurns[self.turnIndex] == false) {
                    self.tapTimes.append(NSDate())
                    
                    self.didTapForTurns[self.turnIndex] = true
                    self.hits++;
                    
                    self.setTitle("Hits: \(self.hits)")
                }
            } else {
                self.tapTimes.append(NSDate())
            }
        }
    }

    func showGame() {
        self.gameStartsAtLabel.setHidden(true)
        self.startTimeLabel.setHidden(true)
        self.button.setColor(UIColor.greenColor())
        self.isPlaying = true
    }

    func startGameTimerDidFire(timer: NSTimer) {
        self.showGame()
    
        self.playTurn()
    
        self.timer = NSTimer.scheduledTimerWithTimeInterval(kTurnSeconds, target:self, selector:Selector("playGameTimerDidFire:"), userInfo:nil, repeats:true)
    }

    func playGameTimerDidFire(timer: NSTimer) {
        self.turnIndex++;
        
        if (self.turnIndex == self.gameLogic.turns.count) {
            timer.invalidate()
            self.timer = nil;
            self.endGame()
        } else {
            self.playTurn()
        }
    }

    func playTurn() {
        let turn = self.gameLogic.turns[self.turnIndex]
        
        let shape = turn.shapeForPlayer(kPlayerIndex);
        let pattern = turn.patternForPlayer(kPlayerIndex)
        
        self.button.setBackgroundImageNamed("shape\(shape)pattern\(pattern)")
    }

    func endGame() {
        self.button.setHidden(true)
        self.gameOverLabel.setAlpha(1.0)
        
        let libraryURL = NSFileManager.defaultManager().URLForDirectory(NSSearchPathDirectory.ApplicationDirectory, inDomain:NSSearchPathDomainMask.UserDomainMask, appropriateForURL:nil, create:true, error:nil)!
        let fileURL = libraryURL.URLByAppendingPathComponent("Taps.plist")
        
        var error: NSError?
        
        let data = NSPropertyListSerialization.dataWithPropertyList(self.tapTimes, format:NSPropertyListFormat.XMLFormat_v1_0, options:Int(0), error:&error)!
        
        let result = data.writeToURL(fileURL, options:NSDataWritingOptions(0), error:&error)
        if (!result) {
            println("(\result)")
        }
    }
}
